clang/test/SemaSYCL/unnamed-kernel.cpp

```diff
@@ -0,0 +1,48 @@
+// RUN: %clang -I %S/Inputs -fsycl-device-only -Xclang -fsycl-int-header=%t.h -fsyntax-only %s -Xclang -verify %s
+// RUN: %clang -I %S/Inputs -fsycl-device-only -Xclang -fsycl-int-header=%t.h -fsycl-unnamed-lambda -fsyntax-only %s -Xclang -verify %s
+#include <sycl.hpp>
+
+namespace namespace1 {
+  template <typename T> class KernelName;
+}
+
+struct MyWrapper {
+private:
+  class InvalidKernelName0 {};
+
+public:
+  void test() {
+    cl::sycl::queue q;
+    // expected-error@+1 {{kernel needs to have a globally-visible name}}
+    class InvalidKernelName1 {};
+    q.submit([&](cl::sycl::handler &h) {
+      h.single_task<InvalidKernelName1>([] {});
+    });
+
+    // expected-error@+1 {{kernel needs to have a globally-visible name}}
+    class InvalidKernelName2 {};
+    q.submit([&](cl::sycl::handler &h) {
+      h.single_task<namespace1::KernelName<InvalidKernelName2>>([] {});
+    });
+
+    // expected-error@11 {{kernel needs to have a globally-visible name}}
+    q.submit([&](cl::sycl::handler &h) {
+      h.single_task<InvalidKernelName0>([] {});
+    });
+
+    // TODO: should this be disallowed or not?
```

Comment:
I do not see why it should be allowed if your diagnostic tells `needs to have a globally-visible name`.

Reply:
It turned out that diagnostic weren't emitted because faulty declaration is the same as for the previous test case: due to the logic in SemaSYCL, we emit such errors per-declaration and not per-enqueue.

I've updated the test case to actually check that this case is also discarded by the compiler